### PR TITLE
Mark the ADBC driver experimental, fix stream error-code + NULL-param bugs, correct comments

### PR DESCRIPTION
**Pull request by ShawnChen-Sirius**
Follow-up to the ADBC driver PR (#132). Two parts:

### Docs / comments (no behavior change)
- Marks the ADBC driver experimental/preview (`bindings.md`, `programs/local/adbc/README.md`).
- Corrects stale stream-concurrency comments to the implemented reject/reuse semantics and drops the reference to the renamed `invalidateActiveStream` (now `reclaimActiveStream`).
- Fixes the executemany/parameter and all-null-column comments.
- `chdb.h`: connection note corrected to "one storage path per process, multiple connections allowed".

### Bug fixes (surfaced by comparing against the official ClickHouse ADBC driver v0.1.0)
- **Streaming-path error codes**: errors on the streaming SELECT path were reported as a blind `INTERNAL`, while the same engine error on the ingest/materialized paths mapped to a precise ADBC code (e.g. `SELECT * FROM missing` gave INTERNAL but ingest into a missing table gave NOT_FOUND). Route the streaming path through `statusForEngineError` so it matches (NOT_FOUND / INVALID_ARGUMENT / …).
- **Single NULL parameter**: `SELECT ?` bound to `None` failed with "first fetch failed" — the single-row path emitted a bare literal NULL (typed `Nothing`, no Arrow output). NULL-carrying columns now always bind as `Nullable(T)` + `\N`; the all-null column uses `String` as its base type so the Nullable wrap doesn't double up.

Regression tests added: `test_streaming_query_error_maps_status_code`, `test_single_null_parameter_selects_null`. Local: unit 59/59 (both artifacts) + validation 215/215 green.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

> [!NOTE]
> ### Fix ADBC driver NULL-param and stream error-code bugs and mark it experimental
> - Marks the ADBC driver as experimental in [bindings.md](https://github.com/chdb-io/chdb-core/pull/145/files#diff-a3bf45cb1d72dd07424e8432a210f45f96447e90c938ec37e7accaf2f7efcfe3) and [README.md](https://github.com/chdb-io/chdb-core/pull/145/files#diff-f3a4c0a0c61d32cecbf51df146e1c595a0221fde04e4658d8000e061086d6369), noting behavior and packaging may change.
> - Fixes parameterized SELECTs with NULL values: columns with any NULLs now bind as `Nullable(<type>)` using `\N` instead of emitting literal NULLs that produced untyped `Nothing` columns.
> - Fixes streaming and query error codes: `statusForEngineError` now parses the terminal `(CLASS_NAME)` token so errors map to the correct ADBC status (e.g. `NOT_FOUND`, `INVALID_ARGUMENT`) instead of always returning `INTERNAL`.
> - Adds `file:` and `chdb:` URI scheme support to `chdbDatabaseSetOption`, including in-memory sentinels and RFC 8089 authority validation; rejects non-localhost authorities with `INVALID_ARGUMENT`.
> - Adds `dropIllegalNullables` to strip forbidden `Nullable` wrappers over composite types (`Tuple`, `Array`, `Map`) when converting Arrow schemas, fixing the `ArrowStream` table function.
>
> 
> #### Changes since #145 opened
>
> - Modified worker-mode process termination in `tests.run_all` module to use `os._exit` with explicit stdout/stderr flushing [42e2f25]
> - Added conditional memory cleanup to `tests.test_adbc_driver.TestAdbcUri._connect_uri` helper method [42e2f25]
> 
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 9adb066.</sup>
> 
>
>

### CODE REVIEW

**Comment by ShawnChen-Sirius** [comment]
@chibugai please review it

File: programs/local/chdb-adbc.cpp [hunk 1]
```diff
@@ -2244,7 +2297,9 @@ AdbcStatusCode executeStreamingSelect(
         /// before executing them, so falling back re-executes nothing.
         if (message.find(CHDB::kErrorStreamingNotSupportedPrefix) != std::string::npos)
             return executeMaterializedSelect(impl, out, rows_affected, error);
-        return setError(error, ADBC_STATUS_INTERNAL, "[chdb] " + message);
+        /// Map the engine error class (UNKNOWN_TABLE, SYNTAX_ERROR, …) to the
+        /// ADBC status code, same as the non-streaming paths.
+        return setError(error, statusForEngineError(message), "[chdb] " + message);
```
Comment (chibugai) [inline programs/local/chdb-adbc.cpp:2317]: Inconsistent error mapping: this change maps engine error classes to ADBC status codes on the plain streaming path (and in wireStreamingResult), but the parameterized streaming path in executeBoundQuery (~line 2041, `return setError(error, ADBC_STATUS_INTERNAL, "[chdb] " + message);` after the `kErrorStreamingNotSupportedPrefix` check) still returns a blind INTERNAL. A parameterized `SELECT * FROM missing_table` will report INTERNAL while the same unparameterized query reports NOT_FOUND. Consider applying `statusForEngineError(message)` there as well for consistency.

File: programs/local/chdb-adbc.cpp [hunk 2]
```diff
@@ -817,26 +847,54 @@ AdbcStatusCode chdbDatabaseSetOption(
...
-                const auto slash = rest.find('/');
-                const std::string authority = slash == std::string::npos ? rest : rest.substr(0, slash);
-                if (!authority.empty() && authority != "localhost")
-                    return setError(
-                        error, ADBC_STATUS_INVALID_ARGUMENT,
-                        "[chdb] unsupported file URI authority '" + authority + "'");
-                rest = slash == std::string::npos ? "" : rest.substr(slash);
+                const std::string auth_rest = after.substr(2);
+                if (auth_rest.find('/') == std::string::npos
+                    && (auth_rest == ":memory:" || auth_rest == "memory"))
```
Comment (chibugai) [inline programs/local/chdb-adbc.cpp]: Percent-decoding inconsistency between sentinel positions: the path-position sentinel check below deliberately decodes first ("Decode first so a percent-encoded ':memory:' still matches"), but this authority-position check compares `auth_rest` raw. So `chdb://%3Amemory%3A` won't match the sentinel and instead falls through to parseFileLikeUri, failing with a confusing "unsupported chdb URI authority '%3Amemory%3A'" error. Consider decoding `auth_rest` before the sentinel comparison for consistent behavior.

```suggestion
                const std::string auth_rest = percentDecode(after.substr(2));
                if (auth_rest.find('/') == std::string::npos
                    && (auth_rest == ":memory:" || auth_rest == "memory"))
```

File: tests/test_adbc_driver.py [hunk 3]
```diff
@@ -340,6 +340,39 @@ def test_ingest_roundtrip_values(self):
             got.column("name").to_pylist(), table.column("name").to_pylist()
         )
 
+    def test_ingest_struct_column(self):
+        # Arrow fields are nullable by default; the schema conversion must not
+        # produce the illegal Nullable(Tuple(...)) for a struct column.
```
Comment (chibugai) [inline tests/test_adbc_driver.py:422]: Test coverage gap: the implementation (dropIllegalNullables in ArrowSchema.cpp) silently strips the Nullable wrapper from struct columns, but neither struct test includes an actual NULL struct value (e.g. `[{"a": 1, "b": "x"}, None]`). Since Arrow structs are nullable by default and nullability is now dropped, ingesting data that genuinely contains null structs may error out or silently coerce to default tuples — this boundary condition should be pinned down by a test so the behavior is explicit rather than accidental.

File: tests/test_adbc_driver.py [hunk 4]
```diff
@@ -780,6 +839,77 @@ def test_non_file_scheme_rejected(self):
...
+        finally:
+            os.chdir(prev)
+            shutil.rmtree(base, ignore_errors=True)
+
+    def _assert_in_memory(self, uri, forbidden_dirname):
+        # In-memory must work AND must not create a dir named after the token.
+        prev = os.getcwd()
+        base = tempfile.mkdtemp(prefix="chdb_mem_")
+        try:
+            os.chdir(base)
```
Comment (chibugai) [inline tests/test_adbc_driver.py]: os.chdir() mutates process-global state, and this pattern is repeated across three new tests (_assert_in_memory, test_chdb_relative_path, test_chdb_relative_named_memory_escape_hatch). If the suite is ever run with parallel execution (e.g. pytest-xdist) these tests can corrupt each other's cwd, and any failure inside the with-block leaves subsequent assertions running from an unexpected directory. Consider extracting a small context manager (e.g. contextlib.chdir on Python 3.11+, or a helper yielding the temp dir) to centralize the chdir/restore/rmtree logic instead of repeating the try/finally in each test.

```suggestion
        # e.g. with _pushd(tempfile.mkdtemp(prefix="chdb_mem_")) as base:
        prev = os.getcwd()
        base = tempfile.mkdtemp(prefix="chdb_mem_")
        try:
            os.chdir(base)
```

File: tests/test_adbc_driver.py [hunk 5]
```diff
@@ -726,6 +767,24 @@ def test_get_table_schema_missing_table_is_not_found(self):
...
+            for sql, want in [
+                ("SELECT * FROM definitely_missing_table", AdbcStatusCode.NOT_FOUND),
+                ("SELEC 1", AdbcStatusCode.INVALID_ARGUMENT),
+                ("SELECT cast('abc' AS Int64)", AdbcStatusCode.INVALID_ARGUMENT),
+            ]:
+                with conn.cursor() as cur:
+                    with self.assertRaises(Exception) as ctx:
+                        cur.execute(sql)
+                        cur.fetchall()
+                    self.assertEqual(ctx.exception.status_code, want, sql)
```
Comment (chibugai) [inline tests/test_adbc_driver.py:884]: assertRaises(Exception) is overly broad, and combining execute() + fetchall() in one block makes the failure mode ambiguous: if the raised exception is not an ADBC error (e.g. an unrelated TypeError), the subsequent `ctx.exception.status_code` access fails with an AttributeError that obscures the real cause. Prefer asserting on the driver's error type (e.g. `adbc_dbapi.Error` / `dbapi.OperationalError`), consistent with checking a status_code afterwards.

```suggestion
                with conn.cursor() as cur:
                    with self.assertRaises(adbc_dbapi.Error) as ctx:
                        cur.execute(sql)
                        cur.fetchall()
                    self.assertEqual(ctx.exception.status_code, want, sql)
```

File: programs/local/chdb-adbc.cpp (hunk 1 above)
Comment (ShawnChen-Sirius) [reply programs/local/chdb-adbc.cpp:2317]: Fixed in 0a43afe — the parameterized streaming path now routes engine errors through `statusForEngineError` as well; a parameterized `SELECT * FROM missing_table` now reports NOT_FOUND (verified with a direct check).

File: programs/local/chdb-adbc.cpp (hunk 2 above)
Comment (ShawnChen-Sirius) [reply programs/local/chdb-adbc.cpp]: Fixed in 0a43afe — the authority-position sentinel now percent-decodes before comparison (suggestion applied), consistent with the path position. Added `chdb://%3Amemory%3A` to test_chdb_memory_forms.

File: tests/test_adbc_driver.py (hunk 3 above)
Comment (ShawnChen-Sirius) [reply tests/test_adbc_driver.py:422]: Good catch — pinned down in 0a43afe with test_ingest_struct_with_null_value. Measured behavior: ClickHouse cannot represent a NULL Tuple, so the outer validity maps to all-NULL nullable fields (`[{'a':1,'b':'x'}, None]` ingests and reads back as `[{'a':1,'b':'x'}, {'a':None,'b':None}]`) — neither an error nor default-value coercion.

File: tests/test_adbc_driver.py (hunk 4 above)
Comment (ShawnChen-Sirius) [reply tests/test_adbc_driver.py]: Fixed in 0a43afe — extracted a `_pushd_tmpdir` context manager (chdir + restore + rmtree) and reused it in all three tests.

File: tests/test_adbc_driver.py (hunk 5 above)
Comment (ShawnChen-Sirius) [reply tests/test_adbc_driver.py:884]: Fixed in 0a43afe — tightened to `assertRaises(adbc_dbapi.Error)` so a non-ADBC exception fails the assertion directly instead of surfacing as an AttributeError on status_code.

File: programs/local/TableFunctionArrowStream.cpp [hunk 6]
```diff
@@ -105,7 +105,7 @@ ColumnsDescription TableFunctionArrowStream::getActualTableStructure(
     }
 
     NamesAndTypesList names_and_types;
-    CHDB::ArrowSchemaWrapper::convertArrowSchema(schema, names_and_types, context);
+    CHDB::ArrowSchemaWrapper::convertArrowSchema(schema, names_and_types, context, /*drop_illegal_nullables=*/true);
```
[the remark this answers is not in the document]
Comment (ShawnChen-Sirius) [reply programs/local/TableFunctionArrowStream.cpp:108]: Good observation, but the suggested gate doesn't work for the actual call shape: bulk ingest creates tables via `CREATE TABLE ... AS SELECT * FROM arrowstream(...)`, where the table function sits on the SELECT side — resolved through `Context::executeTableFunction` (Context.cpp:2968), which never passes `is_insert_query` (defaults to `false`, ITableFunction.h:89). `is_insert_query=true` only reaches table functions in the `CREATE TABLE t AS tf()` / `INSERT INTO FUNCTION tf()` direct forms, which ingest doesn't use. Gating on it would re-introduce the Code 44 failure this change fixes.

On the semantic point: `arrowstream()` is the internal pipe for chdb_arrow_scan/ADBC ingest rather than a general query surface; stripping degrades an outer struct NULL to all-NULL fields, which matches ClickHouse's "Tuple cannot be NULL" semantics and the ingest read-back behavior (pinned by test_ingest_struct_with_null_value). The general query surface — the Python() table function — keeps the historical schema untouched (see da2d456). If arrowstream() is ever exposed as a general-purpose table function, a per-statement-kind gate would be worth revisiting then.

**Comment by ShawnChen-Sirius** [comment]
@wudidapaopao please review this pr

**Review by wudidapaopao — APPROVED** [review]
LGTM
